modules/zone/zone.go

```diff
@@ -93,82 +56,86 @@ func (s *Lookup) DoZonefileLookup(record *dns.Token) (interface{}, zdns.Status,
-		domain = record.RR.Header().Name
-		domain = strings.Join([]string{prefix, domain}, "")
-	default:
-		return nil, zdns.STATUS_NO_OUTPUT, nil
-	}
-	if strings.Count(record.RR.Header().Name, ".") < 2 {
-		return nil, zdns.STATUS_NO_OUTPUT, nil
-	}
-	if domain[len(domain)-1] == '.' {
-		domain = domain[0 : len(domain)-1]
-	}
-	//Verify this didn't succeed before
-	//Make sure it is in Cachehash- if we added it, proceed
-	//if we didn't add it, listen on the chan
-	//	new bool, true-> return no output, false-> proceed
-	//  closed channel-> return no output
-	proceed := false
-	s.Factory.Factory.CHmu.Lock()
-	tmp, found := s.Factory.Factory.CacheHash.Get(domain)
-	if !found {
-		notify = make(chan bool, 10)
-		s.Factory.Factory.CacheHash.Add(domain, notify)
-		s.Factory.Factory.CHmu.Unlock()
-		proceed = true
-	} else {
-		notify = tmp.(chan bool)
-		s.Factory.Factory.CHmu.Unlock()
-		for unsolved := range notify {
-			if unsolved {
-				proceed = true
-				break
-			} else {
-				notify <- false
-				break
-			}
+	for {
+		switch typ := record.RR.(type) {
+		case *dns.NS:
+			nameserver = strings.ToLower(typ.Ns)
+			domain = record.RR.Header().Name
+			domain = strings.Join([]string{prefix, domain}, "")
+		default:
+			return nil, zdns.STATUS_NO_OUTPUT, nil
 		}
-	}
-	if !proceed {
-		return nil, zdns.STATUS_NO_OUTPUT, nil
-	}
-	//First pass looking for a nameserver we know the IP for
-	var result interface{}
-	var status zdns.Status
-	var err error
-	s.Factory.Factory.GlueLock.RLock()
-	if locations, ok := (*s.Factory.Factory.Glue)[nameserver]; ok {
-		result, status, err = LookupHelper(domain, locations, lookup.(*alookup.Lookup))
-		if status == zdns.STATUS_NOERROR && err == nil {
-			s.Factory.Factory.GlueLock.RUnlock()
-			notify <- false
-			return result, status, err
+		if strings.Count(record.RR.Header().Name, ".") < 2 {
+			return nil, zdns.STATUS_NO_OUTPUT, nil
 		}
-	}
-	s.Factory.Factory.GlueLock.RUnlock()
-	//Second pass performing lookups to find a nameserver
-	s.Factory.Factory.GlueLock.Lock()
-	if _, ok := (*s.Factory.Factory.Glue)[nameserver]; !ok {
-		result, status, err = lookup.(*alookup.Lookup).DoLookup(nameserver[0 : len(nameserver)-1])
-		if status == zdns.STATUS_NOERROR && err == nil {
-			addresses := append(result.(alookup.Result).IPv4Addresses, result.(alookup.Result).IPv6Addresses...)
-			(*s.Factory.Factory.Glue)[nameserver] = addresses
-			result, status, err = LookupHelper(domain, addresses, lookup.(*alookup.Lookup))
+		if domain[len(domain)-1] == '.' {
+			domain = domain[0 : len(domain)-1]
+		}
+		s.Factory.Factory.CHmu.Lock()
+		tmp, found := s.Factory.Factory.Hash.Get(domain)
+		if !found {
+			waiting = []*dns.Token{}
+			s.Factory.Factory.Hash.Add(domain, waiting)
+			s.Factory.Factory.CHmu.Unlock()
+		} else {
+			waiting = tmp.([]*dns.Token)
+			waiting = append(waiting, record)
+			s.Factory.Factory.Hash.Add(domain, waiting)
+			s.Factory.Factory.CHmu.Unlock()
+			return nil, zdns.STATUS_NO_OUTPUT, nil
```

Comment:
This is likely due to not being able to understand this logic, but why is `nil` being returned here?